--!optimize 0
local TEST, DONE, PREFIX = require("@test/utils/spec")(...)

local ecs = require("@test/utils/ecs")

local A = ecs.component()
local B = ecs.component()
local C = ecs.component()
local D = ecs.component()
local E = ecs.component()
local F = ecs.component()
local e1 = ecs.entity()

PREFIX("bulk_add")

TEST("components", function(interface)
	ecs.bulk_add(e1, { A, B, C })
	interface:expect_truthy(ecs.has(e1, A))
	interface:expect_truthy(ecs.has(e1, B))
	interface:expect_truthy(ecs.has(e1, C))
end)

PREFIX("bulk_set")

TEST("components", function(interface)
	ecs.bulk_set(e1, { A, B, C }, { 1, 2, 3 })
	interface:expect_truthy(ecs.get(e1, A) == 1)
	interface:expect_truthy(ecs.get(e1, B) == 2)
	interface:expect_truthy(ecs.get(e1, C) == 3)
end)

TEST("components and tag", function(interface)
	ecs.bulk_set(e1, { A, B, C }, { 1, 2, 3 })
	ecs.bulk_set(e1, { D, E, F }, { 4, nil, 6 })
	interface:expect_equal(ecs.get(e1, A), 1)
	interface:expect_equal(ecs.get(e1, B), 2)
	interface:expect_equal(ecs.get(e1, C), 3)
	interface:expect_equal(ecs.get(e1, D), 4)
	interface:expect_equal(ecs.get(e1, E), nil)
	interface:expect_equal(ecs.get(e1, F), 6)
	interface:expect_truthy(ecs.has(e1, E))
end)

PREFIX("bulk_remove")

TEST("components", function(interface)
	ecs.bulk_set(e1, { A, B, C }, { 1, 2, 3 })
	ecs.bulk_remove(e1, { A, B, C })
	interface:expect_equal(ecs.get(e1, A), nil)
	interface:expect_equal(ecs.get(e1, B), nil)
	interface:expect_equal(ecs.get(e1, C), nil)
	interface:expect_falsy(ecs.has(e1, A))
	interface:expect_falsy(ecs.has(e1, B))
	interface:expect_falsy(ecs.has(e1, C))
end)

TEST("components and tag", function(interface)
	ecs.bulk_set(e1, { A, B, C }, { 1, 2, 3 })
	ecs.bulk_set(e1, { D, E, F }, { 4, nil, 6 })
	ecs.bulk_remove(e1, { A, E, F })
	interface:expect_equal(ecs.get(e1, A), nil)
	interface:expect_equal(ecs.get(e1, B), 2)
	interface:expect_equal(ecs.get(e1, C), 3)
	interface:expect_equal(ecs.get(e1, D), 4)
	interface:expect_equal(ecs.get(e1, E), nil)
	interface:expect_equal(ecs.get(e1, F), nil)
	interface:expect_falsy(ecs.has(e1, A))
	interface:expect_truthy(ecs.has(e1, B))
	interface:expect_truthy(ecs.has(e1, C))
	interface:expect_truthy(ecs.has(e1, D))
	interface:expect_falsy(ecs.has(e1, E))
	interface:expect_falsy(ecs.has(e1, F))
end)

return DONE()
